pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -253,43 +255,54 @@ func (ca *cloudCIDRAllocator) updateCIDRAllocation(nodeName string) error {
 		return fmt.Errorf("node %s doesn't have providerID", nodeName)
 	}
 
-	cidrs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
+	ipv6DualStack := utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack)
+	podCIDRs, err := ca.cloud.AliasRangesByProviderID(node.Spec.ProviderID)
 	if err != nil {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
-	if len(cidrs) == 0 {
+	if len(podCIDRs) == 0 {
 		nodeutil.RecordNodeStatusChange(ca.recorder, node, "CIDRNotAvailable")
 		return fmt.Errorf("failed to allocate cidr: Node %v has no CIDRs", node.Name)
 	}
-	_, cidr, err := net.ParseCIDR(cidrs[0])
-	if err != nil {
-		return fmt.Errorf("failed to parse string '%s' as a CIDR: %v", cidrs[0], err)
+	if _, err := netutils.ParseCIDRs(podCIDRs); err != nil {
+		return err
 	}
-	podCIDR := cidr.String()
 
-	if node.Spec.PodCIDR == podCIDR {
-		klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDR)
+	if !ipv6DualStack && node.Spec.PodCIDR == podCIDRs[0] {
+		klog.V(4).Infof("Node %v already has allocated CIDR %v. It matches the proposed one.", node.Name, podCIDRs[0])
 		// We don't return here, in order to set the NetworkUnavailable condition later below.
+	} else if ipv6DualStack && len(node.Spec.PodCIDRs) == len(podCIDRs) {
+		match := true
+		for idx, cidr := range podCIDRs {
+			if node.Spec.PodCIDRs[idx] != cidr {
+				match = false
+				break
```

Comment:
Don't we need to go into the else condition below and set the new podCIDR in this case?

Reply:
Yes, you are right. Changed the code to handle this case correctly.

Reply:
We could just do this:
if len(node.Spec.PodCIDRs) == len (podCIDRs) {
    match := true
    for idx, cidr := range podCIDRs {
	if node.Spec.PodCIDRs[idx] != cidr {
	    match = false
	    break
        }
    }
    if match { skip = true}
}
This handles both cases